trust-quorum/tests/coordinator.rs

```diff
@@ -0,0 +1,646 @@
+        for envelope in envelopes {
+            let msgs = self.network_msgs.entry(envelope.to).or_default();
+            msgs.push(envelope.msg);
+        }
+    }
+}
+
+// A high-level set of actions to drive the state of a peer forward.
+#[derive(Debug, Arbitrary)]
+pub enum Action {
+    /// For each indexed member deliver an in-flight network msg if there is
+    /// one.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.network_msgs`.
+    ///
+    /// We may deliver more than one message to each member.
+    #[weight(50)]
+    DeliverMsgs(Vec<Index>),
+
+    /// For each indexed member drop an in-flight network msg if there is
+    /// one.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.network_msgs`.
+    ///
+    /// We may drop more than one message to each member, which is why we
+    /// use a Vec. We don't worry about shuffling messsages so they are out of
+    /// order, because we use stream oriented transport.
+    //#[weight(3)]
+    //DropMessages(Vec<Index>),
+
+    /// Fake a reply for a delivered message.
+    ///
+    /// The indexes here are used to index into the `PlatformIds` of
+    /// `test_state.delivered_msgs`.
+    #[weight(45)]
+    Reply(Vec<Index>),
+    // Start a reconfiguration that has a possibility of succeeding
+    // given delivered messages.
+    //
+    // It will never fail a validation check.
+    //#[weight(2)]
+    //StartReconfiguration(GeneratedConfiguration),
+
+    // A tick duration in milliseconds
+    #[weight(50)]
+    Tick(
+        #[strategy(
+            (RETRY_TIMEOUT_MS/4..RETRY_TIMEOUT_MS)
+            .prop_map(|ms| Duration::from_millis(ms))
+         )]
+        Duration,
+    ),
+}
+
+/// Informnation about configurations used at test generation time
+#[derive(Debug, Clone, Arbitrary)]
+pub struct GeneratedConfiguration {
+    /// The members of this configuration
+    ///
+    /// We don't use an `Index` here because we know we have a fixed size
+    /// universe to select from, and we want to guarantee uniqueness. If we
+    /// derived `Ord` on `Index` and stored them in a BTreeSet, we'd get a
+    /// unique set of `Index` values. But when used, the output indexes can
+    /// still be duplicated due to the shift implementation used. Therefore we
+    /// instead just choose from a constrained set of usize values that we can
+    /// use directly as indexes into our fixed size structure for all tests.
+    #[strategy(btree_set(0..=255usize, MIN_CLUSTER_SIZE..=MAX_CLUSTER_SIZE))]
+    pub members: BTreeSet<usize>,
+
+    /// An index is roughly equivalent to a threshold, since a threshold cannot
+    /// exceed the number of members in a configuration. Using the index is
+    /// very efficient as opposed to a `prop_flat_map`. When we calculate the
+    /// threshold from the index we use max(2, Index), since the minimum
+    /// threshold is always 2.
```

Comment:
This does bias the index slightly away from being uniformly at random -- I guess it isn't a huge deal but another way to do it might be `2 + index.index(...)`.

Reply:
I think if we do it that way, we could end up going over the max index. I'm not too worried about this distribution :D